modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassSystemMenu.java

```diff
@@ -169,32 +216,14 @@ private void insertMenu(final Menu parent, final MenuBase mb, int pos) {
                     }
                 }
             }
-        });
-
-        for (MenuItemBase item : items) {
-            if (item instanceof MenuBase) {
-                // submenu
-                addMenu(glassMenu, (MenuBase)item);
-            } else {
-                // menu item
-                addMenuItem(glassMenu, item);
-            }
-        }
-        glassMenu.setPixels(getPixels(mb));
-
-        setMenuBindings(glassMenu, mb);
-
-        if (parent != null) {
-            parent.insert(glassMenu, pos);
-        } else {
-            glassSystemMenuBar.insert(glassMenu, pos);
-        }
+        };
     }
 
-    private void setMenuBindings(final Menu glassMenu, final MenuBase mb) {
-        mb.textProperty().addListener(valueModel -> glassMenu.setTitle(parseText(mb)));
-        mb.disableProperty().addListener(valueModel -> glassMenu.setEnabled(!mb.isDisable()));
-        mb.mnemonicParsingProperty().addListener(valueModel -> glassMenu.setTitle(parseText(mb)));
+
+    protected void setMenuBindings(final Menu glassMenu, final MenuBase mb) {
+        mb.textProperty().when(active).addListener(valueModel -> glassMenu.setTitle(parseText(mb)));
+        mb.disableProperty().when(active).addListener(valueModel -> glassMenu.setEnabled(!mb.isDisable()));
+        mb.mnemonicParsingProperty().when(active).addListener(valueModel -> glassMenu.setTitle(parseText(mb)));
```

Comment:
Since you are not calling `get()` from the `when` resulting observable (to allow for invalidation), these need to use a `ChangeListener` instead of an `InvalidationListener`, otherwise, while `active` doesn't change, any change in those menuBase properties will trigger just one notification. 

Alternative, just replace `addListener` with  `subscribe` (since internally uses a `ChangeListener`).

@hjohn does this make sense to you?

Reply:
This could be a problem normally, but I think in this case you won't be able to get this to produce incorrect results.

`parseText` is accessing the property (and it calls `isMnemonicParsing`), but only if the text is not empty.  It's sort of "safe" as an empty text can't contain a mnemonic anyway, and you'd need to change the text at a minimum first to see the effect of turning mnemonic parsing on/off.

A change listener would be more obviously correct, but it is not strictly needed here.

(not all `subscribe`s use change listener, the one that takes a `Runnable` uses invalidation)

Reply:
There is an issue using `when()`, because you add the listener to a new property, not to the original property. Therefore, using an invalidation listener doesn't trigger notifications if you changing values more than once to the original property.
Simply run this test:
```
        BooleanProperty b = new SimpleBooleanProperty();
        BooleanProperty active = new SimpleBooleanProperty(true);

        b.addListener(o -> System.out.println("b: " + b.get()));
        b.when(active).addListener(o -> System.out.println("when: " + b.get()));
        b.set(true);
        b.set(false);
```

It prints:

```
b: true
when: true
b: false
```

so with `when` we miss the changes! (I can go into details of why this happens, but just notice we have a new property after all, and not calling `when.get()` it doesn't get validated anymore).

This can happen in this SystemMenuBar case with the disabled property of any given `Menu` or `MenuItem`, if you simply change it twice, while the application is active. 

My proposal of using `subscribe` implied using the `Consumer` (therefore the "replacement").

Reply:
Thanks, you are right.  The property generated by `when` is not getting revalidated, because its specific `getValue` is never called when using an invalidation listener on it. This seems to be a bit of pitfall that applies to all such mapped properties (not exclusive to `when`).

I'm fine with any of the options.  Option 3 certainly is the most convenient.

I'm more wondering if there should be some action taken in for the `map`, `flatMap`, `when`, `orElse` constructs to make them work better with invalidation listeners, or if that's not possible, a not explaining that adding an invalidation listener to such a construct probably is not doing what you think it is doing.